dali/test/python/test_utils.py

```diff
@@ -435,3 +435,98 @@ def np_type_to_dali(type):
         np.float64: types.FLOAT64,
     }
     return np_types_to_dali_dict[type]
+
+
+class AverageMeter(object):
+    """Computes and stores the average and current value"""
+    def __init__(self):
+        self.reset()
+
+    def reset(self):
+        self.val = 0
+        self.avg = 0
+        self.sum = 0
+        self.count = 0
+        self.avg_last_n = 0
+        self.max_val = 0
+
+    def update(self, val, n=1):
+        self.val = val
+        self.max_val = max(self.max_val, val)
+        self.sum += val * n
+        self.count += n
+        self.avg = self.sum / self.count
+
+
+class FilesDataSet(object):
+
+    def __init__(self, data_path):
+        files = self.get_files_list(data_path)
+        self.data_path = data_path
+        self.classes_no = len(files)
+        self.files_map = {}
+
+        counter = 0
+        for class_no, (_, samples) in enumerate(files):
+            for sample in samples:
+                self.files_map[counter] = (sample, class_no)
+                counter += 1
+        self.size = counter
+
+    @classmethod
+    def get_files_list(cls, data_path):
+        dirs = [
+            (dir_name, [
+                file_path for file_path
+                in map(lambda name: os.path.join(dir_path, name), os.listdir(dir_path))
+                if os.path.isfile(file_path)
+            ])
+            for (dir_name, dir_path)
+            in map(lambda name: (name, os.path.join(data_path, name)), os.listdir(data_path))
+            if os.path.isdir(dir_path)
+        ]
+        dirs.sort(key=lambda dir_files: dir_files[0])
+        for _, files in dirs:
+            files.sort()
+        return dirs
+
+    def __len__(self):
+        return self.size
+
+    def __getitem__(self, i):
+        if i >= self.size:
+            raise StopIteration
+        return self.files_map[i]
+
+
+class SampleLoader(object):
+
+    DATA_SET = FilesDataSet
+
+    def __init__(self, data_path, eager=False):
+        self.data_path = data_path
+        self._data_set = None
+        # If there's a state in the dataset that should be shared between workers
+        # you might want to initialize it eagerly
+        if eager:
+            self._data_set = self.create_data_set()
+
+    def __getattr__(self, key):
+        if key == 'data_set':
+            import_numpy()
+            self._data_set = self._data_set or self.create_data_set()
+            self.__dict__['data_set'] = self._data_set
```

Comment:
Do we need that, I guess __getattr__ will always be called aour code will handle this atribute.

Reply:
It's trickery but I thought it would be acceptable in the test_utils. I need on the one hand eagerly evaluated dataset as some attribute before it is passed to the worker process because dataset instance computes random permutation beforehand and it must be the same across workers, on the other I want python to call __getatrr__ once with the first access to data_set attribute so that it imports numpy in the worker.

By the way it is a separate issue to maybe provide some out of the box handling of dataset permutation in the extarnalsource, so that workers don't have to synchronize between epochos or struggle to agree permutations beforehand., but I got the suggestion to leave it as a folloup.

Reply:
As to the numpy problem here - maybe externalsource should accept optional initialization_callback that setups worker before it starts processing batches?

Reply:
Numpy - no idea.
The rest sounds good.